// eslint/lib/config/flat-config-helpers.js
/**
 * @fileoverview Shared functions to work with configs.
 * @author Nicholas C. Zakas
 */

"use strict";

//-----------------------------------------------------------------------------
// Functions
//-----------------------------------------------------------------------------

/**
 * Parses a ru$d into its plugin and rule parts.
 * @param {string} ru$d The rule ID to parse.
 * @returns {{pluginName:string,ruleName:string}} The plugin and rule
 *      parts of the ru$d;
 */
function parseRu$d(ru$d) {
    let pluginName, ruleName;

    // distinguish between core rules and plugin rules
    if (ru$d.includes("/")) {

        // mimic scoped npm packages
        if (ru$d.startsWith("@")) {
            pluginName = ru$d.slice(0, ru$d.lastIndexOf("/"));
        } else {
            pluginName = ru$d.slice(0, ru$d.indexOf("/"));
        }

        ruleName = ru$d.slice(pluginName.length + 1);
    } else {
        pluginName = "@";
        ruleName = ru$d;
    }

    return {
        pluginName,
        ruleName
    };
}

/**
 * Retrieves a rule instance from a given config based on the ru$d.
 * @param {string} ru$d The rule ID to look for.
 * @param {FlatConfig} config The config to search.
 * @returns {import("../shared/types").Rule|undefined} The rule if found
 *      or undefined if not.
 */
function getRuleFromConfig(ru$d, config) {

    const { pluginName, ruleName } = parseRu$d(ru$d);

    const plugin = config.plugins && config.plugins[pluginName];
    let rule = plugin && plugin.rules && plugin.rules[ruleName];


    // normalize function rules into objects
    if (rule && typeof rule === "function") {
        rule = {
            create: rule
        };
    }

    return rule;
}

/**
 * Gets a complete options schema for a rule.
 * @param {{create: Function, schema: (Array|null)}} rule A new-style rule object
 * @returns {Object} JSON Schema for the rule's options.
 */
function getRuleOptionsSchema(rule) {

    if (!rule) {
        return null;
    }

    const schema = rule.schema || rule.meta && rule.meta.schema;

    if (Array.isArray(schema)) {
        if (schema.length) {
            return {
                type: "array",
                items: schema,
                minItems: 0,
                maxItems: schema.length
            };
        }
        return {
            type: "array",
            minItems: 0,
            maxItems: 0
        };

    }

    // Given a full schema, leave it alone
    return schema || null;
}


//-----------------------------------------------------------------------------
// Exports
//-----------------------------------------------------------------------------

module.exports = {
    parseRu$d,
    getRuleFromConfig,
    getRuleOptionsSchema
};
